tests/SpecTests/TransactionsSpecTest.php

```diff
@@ -0,0 +1,236 @@
+
+        if (isset($expectedCommand['readConcern']['afterClusterTime']) && $expectedCommand['readConcern']['afterClusterTime'] === 42) {
+            $this->assertObjectHasAttribute('readConcern', $actualCommand);
+            $this->assertInternalType('object', $actualCommand->readConcern);
+            $this->assertObjectHasAttribute('afterClusterTime', $actualCommand->readConcern);
+            $this->assertInstanceOf(Timestamp::class, $actualCommand->readConcern->afterClusterTime);
+            unset($expectedCommand['readConcern']['afterClusterTime']);
+
+            /* If "afterClusterTime" was the only assertion for "readConcern",
+             * unset the field to avoid expecting an empty document later. */
+            if (empty($expectedCommand['readConcern'])) {
+                unset($expectedCommand['readConcern']);
+            }
+        }
+
+        /* TODO: Determine if forcing a new libmongoc client in Context is
+         * preferable to skipping the txnNumber assertion. */
+        //unset($expectedCommand['txnNumber']);
+
+        foreach ($expectedCommand as $key => $value) {
+            if ($value === null) {
+                $this->assertObjectNotHasAttribute($key, $actualCommand);
+                unset($expectedCommand[$key]);
+            }
+        }
+
+        $this->assertMatchesDocument($expectedCommand, $actualCommand);
+    }
+
+    /**
+     * Execute an individual test case from the specification.
+     *
+     * @dataProvider provideTests
+     * @param string $name           Test name
+     * @param array  $test           Individual "tests[]" document
+     * @param array  $runOn          Server requirements for the test
+     * @param array  $data           Data fixtures to initialize collection under test
+     * @param string $databaseName   Name of database under test
+     * @param string $collectionName Name of collection under test
+     */
+    public function testTransactions($name, array $test, array $runOn = null, array $data, $databaseName = null, $collectionName = null)
+    {
+        $this->setName($name);
+
+        // TODO: Revise this once a test environment with multiple mongos nodes is available (see: PHPLIB-430)
+        if (isset($test['useMultipleMongoses']) && $test['useMultipleMongoses'] && $this->isShardedCluster()) {
+            $this->markTestSkipped('"useMultipleMongoses" is not supported');
+        }
+
+        if (isset($runOn)) {
+            $this->checkServerRequirements($runOn);
+        }
+
+        if (isset($test['skipReason'])) {
+            $this->markTestSkipped($test['skipReason']);
+        }
+
+        $databaseName = isset($databaseName) ? $databaseName : $this->getDatabaseName();
+        $collectionName = isset($collectionName) ? $collectionName : $this->getCollectionName();
+
+        $context = Context::fromTransactions($test, $databaseName, $collectionName);
+        $this->setContext($context);
+
+        $this->dropTestAndOutcomeCollections();
```

Comment:
According to the spec, it looks like the following should occur:
```
(3) Create a collection object from the MongoClient,
using the database_name and collection_name fields of the YAML file.

(4) Drop the test collection, using writeConcern "majority".

(5) Execute the "create" command to recreate the collection, using writeConcern "majority".
(Creating the collection inside a transaction is prohibited, so create it explicitly.)
```

It looks like this covers steps 4 and 5, but where is step 3 done?

Reply:
`dropTestAndOutcomeCollections()` internally creates a collection by calling `Context::getCollection()`. This is identical to how we create a collection for each operation, with the notable difference that the operation may have options to provide.

Since collection _and_ database options vary by each operation, I saw no reason to create the objects up front as we do with the Client and both Session objects.

Later on in the transactions spec README, we have:

> **11.4**: Create a Collection object from the Database, using the collection_name field at the top level of the test file. If collectionOptions is present create the Collection object with the provided options. Otherwise create the object with the default options.